test/AR_CLR_KLM_tests.jl

```diff
@@ -0,0 +1,359 @@
+        end
+        Γ[j,:,:] ./= n
+        D[:,j] = G[1,:,j] - Γ[j,:,:]*inv(Ω)*gn
+        end
+        return(n,k,p,gn, Ω, D, P)
+    end
+end
+
+"""
+    klm(gi::Function)
+
+Return test statistic for KLM method.
+"""
+
+function klm(gi::Function)
+    SP = statparts(gi)
+    function(θ)
+        (n,k,p,gn, Ω, D, P) = SP(θ)
+        return n*(gn'*Ω^(-1/2)*P(Ω^(-1/2)*D)*Ω^(-1/2)*gn)[1]
+    end
+end
+
+
+"""
+    clr(gi::Function)
+
+Return test statistic for CLR method.
+"""
+function clr(gi::Function)
+  SP = statparts(gi)
+  function(θ)
+        (n,k,p,gn, Ω, D, P) = SP(θ)
+        rk = eigmin(n*D'*inv(Ω)*D)
+        AR  = (n*gn'*inv(Ω)*gn)[1]
+        lm = (n*(gn'*Ω^(-1/2)*P(Ω^(-1/2)*D)*Ω^(-1/2)*gn))[1]  
+        lr = 1/2*(AR - rk + sqrt( (AR-rk)^2 + 4*lm*rk))
+
+        # simulate to find p-value
+        S = 5000
+        function randc(k,p,r,S)
+        χp = rand(Chisq(p),S)
+        χkp = rand(Chisq(k-p),S)
+        0.5.*(χp .+ χkp .- r .+
+              sqrt.((χp .+ χkp .- r).^2 .+ 4 .* χp.*r))
+        end
+        csim = randc(k,p,rk,S)
+        return mean(csim.<=lr)
+    end
+end
+
+
+
+function plot_cr(β,V, tests::AbstractArray{Function}, labels; ngrid=30)
+  lb = β - sqrt.(diag(V))*5
+  ub = β + sqrt.(diag(V))*5
+  fig=scatter([β0[1]], [β0[2]], markersize=8, legend=false,
+              xlabel=L"\beta_1", ylabel=L"\beta_2")
+  ntest = 1000
+  βtest = [rand(2).*(ub-lb) .+ lb for i in 1:ntest]
+  pval = tests[1].(βtest)
+  βtest = vcat(βtest'...)
+  crit = 0.9
+  fig=scatter!(βtest[:,1],βtest[:,2], group=(pval.<crit), legend=false,
+               markersize=4, markerstrokewidth=0.0, seriesalpha=0.5,
+               palette=:heat)
+  b1 = lb[1]:(ub[1]-lb[1])/ngrid:ub[1]
+  b2 = lb[2]:(ub[2]-lb[2])/ngrid:ub[2]
+  colors = [:black, :red, :blue, :green]
+  for t in 1:length(tests)
+    fig=contour!(b1,b2,(a,b)->tests[t]([a,b]),
+             levels = [0.9, 0.95],
+             contour_labels=false, legend=false,
+             label = labels[t],
+             c=cgrad([colors[t],colors[t]],[0.0,1.0]))
+  end
+  fig
+end
+
+
+
+function gmmVar(θ,gi,W)
+  g = gi(θ)
+  n = size(g)[1]
+  D = ForwardDiff.jacobian(θ->mean(gi(θ),dims=1),θ)
+  Σ = cov(gi(θ))
+  1/n*inv(D'*W*D)*(D'*W*Σ*W*D)*inv(D'*W*D)
+end
+function gmmObj(θ,gi,W)
+  g = gi(θ)
+  m = mean(g,dims=1)
+  (size(g)[1]*( m*W*m')[1]) # return scalar, not 1x1 array
+end
+
+
+
+n = 50
+k = 2
+iv =3 
+π0 = vcat(0.1*diagm(0=>ones(k)),0.2*ones(iv-k,k)) 
+ρ = 0.5  
+data = IVLogitShare(n,β0,π0,ρ)
+opt1 = optimize(θ->gmmObj(θ, get_gi(data) ,I),
+                β0, BFGS(), autodiff =:forward)
+β1 = opt1.minimizer
+V1 = gmmVar(β1,get_gi(data),I)
+  
+KLM_f = klm(get_gi(data))
+pklm = θ -> cdf(Chisq(length(β1)), KLM_f(θ) )
+
+AR_f = AR_gmm_obj(get_gi(data))
+par  = θ ->cdf(Chisq(size(data.z)[2]), AR_f(θ) ) 
+pclr  = clr(get_gi(data))
+pwald = θ -> cdf(Chisq(length(β1)),(θ-β1)'*inv(V1)*(θ-β1))
+plot_cr(β1,V1, [pclr, pklm, par, pwald],
+        ["CLR","KLM","AR","Wald"], ngrid=40)
+
+
```

Comment:
Comparing output to existing code is a fine method for manual testing while developing, but it's hard to automate and doesn't work for completely new code. I think a better test would be to check that simulated p-values are approximately uniform. The notes do this visually, but an automated test of it could be added here.